xc/device_common.ml

```diff
@@ -338,22 +339,87 @@ let is_upstream_qemu domid =
     with_xs (fun xs -> xs.Xs.read (sprintf "/libxl/%d/dm-version" domid)) = "qemu_xen"
   with _ -> false
 
-let qmp_write_and_read domid ?(read_result=true) cmd  =
-  try
-    let open Qmp in
-    debug "QMP.Command for domid %d: %s" domid (string_of_message cmd);
-    let c = Qmp_protocol.connect (qmp_libxl_path domid) in
-    finally
-      (fun () ->
-         Qmp_protocol.negotiate c;
-         Qmp_protocol.write c cmd;
-         if read_result then
-           Some (Qmp_protocol.read c)
-         else None
-      )
-      (fun () -> Qmp_protocol.close c)
-  with e ->
-    error "Caught exception attempting to write qmp message: %s" (Printexc.to_string e);
-    None
+(** [make_id_generator] returns a function that creates unique IDs using
+ * an internal counter *)
+let make_id_generator name =
+  let count = ref 0 in
+  fun domid -> incr count; Printf.sprintf "%s-%06d-%d" name !count domid
+
+let make_qmp_id = make_id_generator "qmp"
+
+(** [qmp_send_cmd_internal connection domid cmd] sends [cmd] to QEMU, waits
+ * for the *matching* response, and returns the result. It uses an
+ * already open and negotiated [connection]. Commands are tagged with an
+ * identifier and the function waits for the matching response, skipping
+ * all other responses it receives while waiting.
+*)
+let qmp_send_cmd_internal connection domid cmd =
+  let id      = make_qmp_id domid in
+  let msg     = Qmp.Command(Some id, cmd) in
+  let msg'    = Qmp.string_of_message msg in
+  debug "QMP command for domid %d: %s" domid msg';
+  let rec wait_for_result id =
+    match Qmp_protocol.read connection with
+    (* no ID *)
+    | Qmp.Greeting(_)
+    | Qmp.Command(_, _)
+    | Qmp.Error(None, _)
+    | Qmp.Success(None, _)
+    | Qmp.Event(_) as resp ->
+      let resp' = Qmp.string_of_message resp in
+      debug "%s: skipping QMP message while waiting for %s: %s"
+        __LOC__ id resp';
+      wait_for_result id
+
+    (* wrong ID *)
+    | Qmp.Success(Some id',_) as resp when id <> id' ->
+      let resp' = Qmp.string_of_message resp in
```

Comment:
I was wondering what would happen if we had racing threads, that potentially could read the id of other threads here and discard their reponses, blocking the other threads forever. 

@idruzhinin indicated that the qmp monitor prevents this race by using the negotiating step. At any time, only one thread will ever be past Qmp_protocol.negotiate, so any wrong response id seen at this point will belong to a command for which we already closed the fd previously and therefore we don't care about its response anymore, so it's right to log and ignore wrong ids as done here.

However, there's one thing that the Qmp_protocol.negotiate step won't protect: incomplete negotiation headers racing from two threads. To prevent the interleaving of the negotiation headers, we need to have some sort of mutex around Qmp_protocol.negotiate for each domid (see below).